audio-worklet/src/bitcrusher.js

```diff
@@ -0,0 +1,111 @@
+   * @param {Number} options.inputChannels the number of input channels
+   * @param {Number} options.outputChannels the number of output channels
+   * @param {Number} options.bitDepth the number of bits for each sample
+   * @param {Number} options.reduction the degree of reduction
+   */
+  constructor (context, options) {
+    if (context == null) throw "No AudioContext given";
+    if (options == null) options = {}; 
+    this.context_ = context;
+    
+    //... Default to maximum bit depth and no reduction and let browser decide
+    //... buffersize unless specified
+    this.bitDepth = options.bitDepth || 24; 
+    this.reduction = options.reduction || 1;
+    let bufferSize = options.bufferSize || 0;
+    let inputChannels = options.inputChannels || 1;
+    let outputChannels = options.inputChannels || 1;
+
+    this.node_ = this.context_.createScriptProcessor(bufferSize, 
+      options.inputChannels, options.outputChannels);
+
+    this.node_.onaudioprocess = this.onaudioprocess_.bind(this); 
+
+    //... Let clients connect to bitcrusher via input and output
+    //... e.g oscillator.connect(bitcrusher.input) 
+    //... bitcrusher.output.connect(context.destination)
+    this.input = new GainNode(this.context_);
+    this.output = new GainNode(this.context_);
+    this.input.connect(this.node_).connect(this.output);
+  }
+  
+  /**
+   * Bit crush upon receiving input audio signal, applying signal distortion 
+   * effects to the output buffer
+   * @param  {AudioProcessingEvent} event holds input and output buffers
+   */
+  onaudioprocess_ (event) {
+    for (let i = 0; i < event.inputBuffer.numberOfChannels; i++) {
+      this.processBuffer_(this.reduction, this.bitDepth, 
+        event.inputBuffer.getChannelData(i), 
+        event.outputBuffer.getChannelData(i));
+    }
+  }
+
+  /**
+   * Reduce the information in given inputBuffer with specified factor and 
+   * precision, placing results in outputBuffer.
+   * @param  {Number} factor the amount of sample rate reduction
+   * @param  {Number} precision the bit depth of post-processed samples   
+   * @param  {Float32Array} inputBuffer pre-processed buffer
+   * @param  {Float32Array} outputBuffer post-processed buffer
+   */
+   processBuffer_ (factor, precision, inputBuffer, outputBuffer) {
+    // Set phaser to exceed 1 every [factor] sample
+    if (factor <= 0) factor = 1; 
+    const phaserInc = 1 / factor;
+    let phaser = 1;
+    const scale = Math.pow(2, precision);
+    
+    // Add new bit crushed sample to outputBuffer at specified interval
+    let last;
+    for (let j = 0; j < inputBuffer.length; j++) {
+      if (phaser >= 1) {
+        // Scale up and round off low order bits
+        let temp = inputBuffer[j] * scale;
+        let rounded = Math.round(temp);
+        last = rounded / scale; 
+        phaser = 0;
+      }
+      phaser += phaserInc;      
+      outputBuffer[j] = last;
+    }
+  }
+}
```

Comment:
Don't you need an internal buffer?  And shouldn't phaser be a "global" defined outside of processBuffer_?

My thinking on this:  Let's say the factor is 3 and inputBuffer size is 4.  The first input sample gets duplicated 3 times.  The 3 input sample should be duplicated 3 times.  But since you only have 4 samples in the input, doesn't your code only duplicate this last sample 1 time?  This doesn't match my thinking of what sample rate reduction should be doing.